### PR TITLE
Add CompilationOutputFilePath to project snapshot

### PR DESCRIPTION
**Pull request by tmat**
EnC needs to know the compilation output path for a given project (IntermediateOutputPath). Currently the path is tracked in VisualStudioWorkspaceImpl, which is not available OOP. Move the path to the project snapshot.

Adds new Workspace public API:

```C#
readonly struct CompilationOutputFilePaths : IEquatable<CompilationOutputFilePaths>
{
   public string? AssemblyPath { get; }
   public CompilationOutputFilePaths WithAssemblyPath(string? path);
}

class Project
{
   public CompilationOutputFilePaths CompilationOutputFilePaths { get; }
}

class ProjectInfo
{
   public CompilationOutputFilePaths CompilationOutputFilePaths { get; }
   public ProjectInfo WithCompilationOutputFilePaths(in Microsoft.CodeAnalysis.CompilationOutputFilePaths paths);
}

class Solution 
{
   public Solution WithProjectCompilationOutputFilePaths(ProjectId projectId, in CompilationOutputFilePaths paths);
}

```

Contributes to #42179

### CODE REVIEW

**Comment by tmat** [comment]
@jasonmalinowski @CyrusNajmabadi PTAL

File: src/EditorFeatures/TestUtilities/TestExtensionErrorHandler.cs [hunk 1]
```diff
@@ -30,7 +31,14 @@ public void HandleError(object sender, Exception exception)
...
-            ExceptionUtilities.FailFast(exception);
+            RaiseFailFastException(IntPtr.Zero, IntPtr.Zero, 0);
+
+            // the RaiseFailFastException above may have not actually killed the process if a
+            // debugger was attached. Fall back to the CLR FailFast.
+            Environment.FailFast("FailFast was issued from devenv.exe.  A crash dump should be available.");
         }
+
+        [DllImport("kernel32")]
+        static extern void RaiseFailFastException(IntPtr exceptionRecord, IntPtr contextRecord, uint flags);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/TestUtilities/TestExtensionErrorHandler.cs]: can we make ths cahnge in a separate PR?  seems not required for this particular piece of work.

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs [hunk 2]
```diff
@@ -39,7 +39,7 @@ internal sealed class EditAndContinueWorkspaceService : IEditAndContinueWorkspac
         private readonly EditAndContinueDiagnosticUpdateSource _emitDiagnosticsUpdateSource;
         private readonly EditSessionTelemetry _editSessionTelemetry;
         private readonly DebuggingSessionTelemetry _debuggingSessionTelemetry;
-        private readonly ICompilationOutputsProviderService _compilationOutputsProvider;
+        private readonly Func<Project, CompilationOutputs> _compilationOutputsProvider;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs]: not a huge fan of passing around a delegate like this.  Doesn't seem to match our general patterns here.  Even if it's a single method interface, that seems pretty clear to me.

File: src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            => AssemblyPath == other.AssemblyPath;
+
+        public override int GetHashCode()
+            => AssemblyPath?.GetHashCode() ?? 0;
+
+        public static bool operator ==(CompilationOutputFilePaths left, CompilationOutputFilePaths right)
+            => left.Equals(right);
+
+        public static bool operator !=(CompilationOutputFilePaths left, CompilationOutputFilePaths right)
+            => !left.Equals(right);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs]:
```suggestion
            => !(left == right);
```

File: src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+using System;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Paths of files produced by the compilation.
+    /// </summary>
+    public readonly struct CompilationOutputFilePaths : IEquatable<CompilationOutputFilePaths>, IObjectWritable
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs]: is a struct a potentil concern in the future.  i know we've tended to avoid them in these APIs because chaging them over time can be harder.

File: src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+        public static bool operator ==(CompilationOutputFilePaths left, CompilationOutputFilePaths right)
+            => left.Equals(right);
+
+        public static bool operator !=(CompilationOutputFilePaths left, CompilationOutputFilePaths right)
+            => !left.Equals(right);
+
+        void IObjectWritable.WriteTo(ObjectWriter writer)
+        {
+            writer.WriteString(AssemblyPath);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs]:
```suggestion
            => writer.WriteString(AssemblyPath);
```

File: src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+#nullable enable
+
+using System;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Paths of files produced by the compilation.
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs]: it's a little odd to me that we have this sturct that indicates it's for multiple file paths.  But we only expose one path, and that path is itself optional.  so we have other paths we think we would expose here in hte future?

**Comment by CyrusNajmabadi** [comment]
I don't see any issues here.  but i'm not a roslyn project-system guru.  Would like @jasonmalinowski to check for any concerns.

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs (hunk 2 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs]: This is simpler than defining an interface and implementing a class, and it's equivalent. We already have `Action<DebuggingSessionTelemetry.Data> _reportTelemetry` right below it.

File: src/EditorFeatures/TestUtilities/TestExtensionErrorHandler.cs (hunk 1 above)
Comment (tmat) [reply src/EditorFeatures/TestUtilities/TestExtensionErrorHandler.cs]: It is required - this removes an ambiguous definition of `ExceptionUtilities` which conflicts with another implementation of `ExceptionUtilities` that I need to use.

File: src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs (hunk 3 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs]: Why?

File: src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs (hunk 4 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs]: Why is it harder to add a field to a struct?

File: src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs (hunk 6 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs]: > so we have other paths we think we would expose here in hte future?

Yes, we expect more paths. See https://github.com/dotnet/roslyn/pull/41745/files/9f60efe2f6f0df7477a0856bca30f88f46e18ba3#diff-74b03a1b69714861b192cbc422ed2660R24-R27

File: src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/CompilationOutputFilePaths.cs]: it's more symmetric.  Specifically, it means that we only need to update one and we get the other for free.  :)

**Comment by jasonmalinowski** [comment]
This is on my review list; should get to it tomorrow.

File: src/EditorFeatures/TestUtilities/TestExtensionErrorHandler.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/EditorFeatures/TestUtilities/TestExtensionErrorHandler.cs]: Can we resolve the conflict in some other way? This implies we're just copy/pasting code then?

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs (hunk 2 above)
Comment (jasonmalinowski) [reply src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs]: Agreed with @CyrusNajmabadi an interface seems appropriate.

File: src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpCompilerOptionsTests.cs [hunk 7]
```diff
@@ -114,32 +114,28 @@ public void ProjectOutputPathAndOutputExeNameChange()
             project.SetOutputFileName(initialPath);
             Assert.Equal(initialPath, project.GetOutputFileName());
 
-            var outputs = (CompilationOutputFilesWithImplicitPdbPath)environment.Workspace.GetCompilationOutputs(project.Test_VisualStudioProject.Id);
-            Assert.Equal(initialPath, outputs.AssemblyFilePath);
+            Assert.Equal(initialPath, project.Test_VisualStudioProject.IntermediateOutputFilePath);
```
Comment (jasonmalinowski) [inline src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpCompilerOptionsTests.cs]: Note the tests here changed in a subtle way: now this is asserting the underlying value of the IntermediateOutputFilePath on VisualStudioProject, but nothing is actually asserting it's getting to the project snapshot itself, I don't think?

File: src/EditorFeatures/TestUtilities/TestExtensionErrorHandler.cs (hunk 1 above)
Comment (tmat) [reply src/EditorFeatures/TestUtilities/TestExtensionErrorHandler.cs]: No, this is only a test helper code.

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs (hunk 2 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs]: I don't think we should need an interface just to implement a test hook.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs [hunk 8]
```diff
@@ -241,18 +241,10 @@ public ParseOptions ParseOptions
...
-                _workspace.SetCompilationOutputs(Id, new CompilationOutputFilesWithImplicitPdbPath(value));
-
-                // Unlike OutputFilePath and OutputRefFilePath, the intermediate output path isn't represented in the workspace anywhere;
-                // thus, we won't mutate the solution. We'll still call ChangeProjectOutputPath so we have the rest of the output path tracking
-                // for any P2P reference conversion.
-                ChangeProjectOutputPath(ref _intermediateOutputFilePath, value, s => s, w => { });
-            }
+            set => ChangeProjectOutputPath(ref _intermediateOutputFilePath,
+                       value,
+                       s => s.WithProjectCompilationOutputFilePaths(Id, new CompilationOutputFilePaths(value)),
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs]: Agreed. Will add With-.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs [hunk 9]
```diff
@@ -241,18 +241,10 @@ public ParseOptions ParseOptions
         internal string IntermediateOutputFilePath
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs]: Sure, I'll follow up with a rename.

File: src/Features/Core/Portable/Emit/CompilationOutputFilesWithImplicitPdbPath.cs [hunk 10]
```diff
@@ -1,15 +1,15 @@
...
 using System.Reflection.PortableExecutable;
 using Microsoft.CodeAnalysis.Debugging;
-using Microsoft.CodeAnalysis.Emit;
 using Roslyn.Utilities;
 
-namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+namespace Microsoft.CodeAnalysis.Emit
 {
     /// <summary>
     /// Provides access to compilation outputs based only on the path of the output asssembly.
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/Features/Core/Portable/Emit/CompilationOutputFilesWithImplicitPdbPath.cs]: > Or is the logic once there's more to consume than just the string then we don't need this type?

That's correct.

File: src/EditorFeatures/Test/EditAndContinue/EditSessionActiveStatementsTests.cs [hunk 11]
```diff
@@ -115,7 +115,7 @@ private sealed class Validator
 
                 var activeStatementProvider = new TestActiveStatementProvider(activeStatements);
                 var mockDebuggeModuleProvider = new Mock<IDebuggeeModuleMetadataProvider>();
-                var mockCompilationOutputsProvider = new MockCompilationOutputsProviderService();
+                var mockCompilationOutputsProvider = new Func<Project, CompilationOutputs>(_ => new MockCompilationOutputs(Guid.NewGuid()));
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/EditorFeatures/Test/EditAndContinue/EditSessionActiveStatementsTests.cs]: Yes.

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs [hunk 12]
```diff
@@ -56,12 +56,12 @@ internal sealed class EditAndContinueWorkspaceService : IEditAndContinueWorkspac
         internal EditAndContinueWorkspaceService(
             Workspace workspace,
             IActiveStatementTrackingService activeStatementTrackingService,
-            ICompilationOutputsProviderService compilationOutputsProvider,
             IDiagnosticAnalyzerService diagnosticService,
             EditAndContinueDiagnosticUpdateSource diagnosticUpdateSource,
             IActiveStatementProvider activeStatementProvider,
             IDebuggeeModuleMetadataProvider debugeeModuleMetadataProvider,
-            Action<DebuggingSessionTelemetry.Data>? reportTelemetry = null)
+            Func<Project, CompilationOutputs>? testCompilationOutputsProvider = null,
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs]: The real compilation outputs require files existing on disk. Some tests do create real files, but most tests don't.

File: src/EditorFeatures/Test/EditAndContinue/EditSessionActiveStatementsTests.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (tmat) [reply src/EditorFeatures/Test/EditAndContinue/EditSessionActiveStatementsTests.cs]: The mock is useful so that most of the tests don't need to create files.

**Comment by tmat** [comment]
@CyrusNajmabadi @jasonmalinowski Rebased and implemented feedback.

File: src/EditorFeatures/TestUtilities/TestExtensionErrorHandler.cs [hunk 13]
```diff
@@ -30,7 +31,14 @@ public void HandleError(object sender, Exception exception)
                 return;
             }
 
-            ExceptionUtilities.FailFast(exception);
+            RaiseFailFastException(IntPtr.Zero, IntPtr.Zero, 0);
+
+            // the RaiseFailFastException above may have not actually killed the process if a
+            // debugger was attached. Fall back to the CLR FailFast.
+            Environment.FailFast("FailFast was issued from devenv.exe.  A crash dump should be available.");
```
Comment (jasonmalinowski) [inline src/EditorFeatures/TestUtilities/TestExtensionErrorHandler.cs:38]: You'll have to merge this with recent changes @jaredpar made. He removed the FailFast since that was making it hard to figure out which test was impacted.

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs [hunk 14]
```diff
@@ -390,6 +399,11 @@ internal sealed class ProjectAttributes : IChecksummedObject, IObjectWritable
             /// </summary>
             public string? OutputRefFilePath { get; }
 
+            /// <summary>
+            /// The path to the compiler output file (module or assembly).
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs:403]: Seems like this comment would change once the struct had more fields? Is there a way to write this that avoids that now?

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs [hunk 15]
```diff
@@ -449,6 +465,7 @@ internal sealed class ProjectAttributes : IChecksummedObject, IObjectWritable
                 Optional<string?> filePath = default,
                 Optional<string?> outputPath = default,
                 Optional<string?> outputRefPath = default,
+                Optional<CompilationOutputFilePaths> compilationOutputPaths = default,
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs:468]: Do we only use the Optional if it's a reference type, more or less? (VersionStamp is different.) I suspect we might not have a constraint on Optional only because it sometimes got used as a hack around not having nullable reference types, but we might have been stricter for public API.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs [hunk 16]
```diff
@@ -244,21 +244,13 @@ public string AssemblyName
...
-                // Unlike OutputFilePath and OutputRefFilePath, the intermediate output path isn't represented in the workspace anywhere;
-                // thus, we won't mutate the solution. We'll still call ChangeProjectOutputPath so we have the rest of the output path tracking
-                // for any P2P reference conversion.
-                ChangeProjectOutputPath(ref _intermediateOutputFilePath, value, s => s);
-            }
+            get => _compilationOutputAssemblyFilePath;
+            set => ChangeProjectOutputPath(
+                       ref _compilationOutputAssemblyFilePath,
+                       value,
+                       s => s.WithProjectCompilationOutputFilePaths(Id, s.State.GetRequiredProjectState(Id).CompilationOutputFilePaths.WithAssemblyPath(value)));
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:253]: Why is this reaching into the state directly rather than reading it via the public API?

File: src/Features/Core/Portable/Emit/CompilationOutputFilesWithImplicitPdbPath.cs [hunk 17]
```diff
@@ -1,15 +1,15 @@
...
 using System;
 using System.IO;
 using System.Linq;
 using System.Reflection.PortableExecutable;
 using Microsoft.CodeAnalysis.Debugging;
-using Microsoft.CodeAnalysis.Emit;
 using Roslyn.Utilities;
 
-namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+namespace Microsoft.CodeAnalysis.Emit
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/Emit/CompilationOutputFilesWithImplicitPdbPath.cs:12]: Why moving this here versus just into EditAndContinue? Does anything else use it at this point? Would prefer we kept it in that namespace until we had a broader need.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectOptionsProcessor.cs [hunk 18]
```diff
@@ -204,11 +204,11 @@ private void UpdateProjectOptions_NoLock()
             _project.AssemblyName = _commandLineArgumentsForCommandLine.CompilationName ?? _project.AssemblyName;
             _project.CompilationOptions = compilationOptions;
 
-            var fullIntermediateOutputPath = _commandLineArgumentsForCommandLine.OutputDirectory != null && _commandLineArgumentsForCommandLine.OutputFileName != null
-                                                    ? Path.Combine(_commandLineArgumentsForCommandLine.OutputDirectory, _commandLineArgumentsForCommandLine.OutputFileName)
-                                                    : _commandLineArgumentsForCommandLine.OutputFileName;
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectOptionsProcessor.cs:209]: The ? and : at the front of the line makes this a lot easier to read, especially on GitHub where the fundamental control structure now requires scrolling.

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs [hunk 19]
```diff
@@ -511,6 +531,7 @@ public void WriteTo(ObjectWriter writer)
                 writer.WriteString(FilePath);
                 writer.WriteString(OutputFilePath);
                 writer.WriteString(OutputRefFilePath);
+                CompilationOutputFilePaths.WriteTo(writer);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs:534]: Do we have any tests for serialization? I don't remember if this means it now works for free or if we also have to update some checksumming stuff.

File: src/Features/Core/Portable/Emit/CompilationOutputFilesWithImplicitPdbPath.cs (hunk 17 above)
Comment (tmat) [reply src/Features/Core/Portable/Emit/CompilationOutputFilesWithImplicitPdbPath.cs:12]: CompilationOutputs are in CodeAnalysis.Emit. It's Emit artifact, not specific to EnC. Even though it's currently used only by EnC.

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs (hunk 19 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs:533]: We do, I have improved them a bit.

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs (hunk 15 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs:467]: There are existing `Optional<bool>` parameters. I don't think it matters if it's `Optional<T>` or `Nullable<T>`. I'm using Optional for consistency. Eventually, we'll just use records for this.
